### PR TITLE
EOS-16878: SSPL resource agent improvements

### PR DESCRIPTION
**Pull request by VijayThakkar132**
Signed-off-by: Vijay Thakkar <vijaykumar.thakkar@seagate.com>

## Problem Statement
<pre>
  <code>
  Story Ref (if any):
    EOS-16878: HA (pcs) did not detect that SSPL had crashed and did not restart it
  </code>
</pre>
## Unit testing on RPM done
<pre>
  <code>
  Yes
  </code>
</pre>
## Problem Description
<pre>
  <code>
    Observed that demoted node is taking too much time after the forcefully kill.
  </code>
</pre>
## Solution
<pre>
  <code>
    Added code changes to remove state file in the stop method and also added couple of improvement
  </code>
</pre>
## Unit Test Cases
<pre>
  <code>
    Tested changes in the live setup and observed that sspl service starts immediately after the demotion.
  </code>
</pre>

### CODE REVIEW

**Review by Madhura-08 — APPROVED** [review]

File: ha/resource/sspl [hunk 1]
```diff
@@ -121,13 +121,21 @@ stateful_start() {
         # CRM Error - Should never happen
         log "START (master)..."
         systemctl start sspl-ll
+        while ! systemctl is-active sspl-ll | grep "active"
```
Comment (ajaysrivas) [inline ha/resource/sspl:124]: you should be checking "failed" status as well and if service fails to start, return proper OCF code.
Applies to code below too. You can add another function as code will be common and call the function from both the places.

**Review by ajaysrivas — CHANGES_REQUESTED** [review]

File: ha/resource/sspl (hunk 1 above)
Comment (VijayThakkar132) [reply ha/resource/sspl]: Addressed comment in the latest patch

File: ha/resource/sspl [hunk 2]
```diff
@@ -115,22 +115,45 @@ stateful_check_state() {
...
 
+start_sspl_service() {
+        systemctl start sspl-ll
+        echo "START (master): $?"
+        while true
+        do
+            status=`systemctl is-active sspl-ll`
+            if [[ $status == "active" ]]; then
+                return 0
+            elif [[ $status == "failed" ]]; then
```
Comment (ajaysrivas) [inline ha/resource/sspl:126]: what about "unknown" status?
How should that be handled?
@ajay-paratmandali, have you checked that in dynamicFid resource agent?

Comment (VijayThakkar132) [reply ha/resource/sspl:126]: @ajaysrivas What is the expected behavior when the current service state is "unknown"?  Do we need to wait till the service starts or returns immediately?

Comment (ajaysrivas) [reply ha/resource/sspl:126]: Not sure. This is the reason, I asked question :-)
Probably we should check if it remains "unknown" for some time (5 sec) then return 1.

Comment (VijayThakkar132) [reply ha/resource/sspl:127]: Yes, It's good to wait for 5 sec. Did the changes in the latest patch.

**Review by indrajitzagade — APPROVED** [review]

File: ha/resource/sspl (hunk 2 above)
Comment (ajaysrivas) [reply ha/resource/sspl:127]: I meant, continuously for 5 sec. So the count should be set to zero in other if conditions. I guess, 5 sec is fine for a stable status, so this is okay too.

File: ha/resource/sspl [hunk 3]
```diff
@@ -115,22 +115,50 @@ stateful_check_state() {
...
         log "START (master)..."
-        systemctl start sspl-ll
-        log "START (master): $?"
-        return $OCF_RUNNING_MASTER
+        start_sspl_service
+        if [ $? == 0 ]; then
+            return $OCF_RUNNING_MASTER
+        elif [ $? == 1 ]; then
+            return $OCF_FAILED_MASTER
+        else
```
Comment (ajaysrivas) [inline ha/resource/sspl:149]: return code is either 0 or 1, it will never reach here.

Comment (VijayThakkar132) [reply ha/resource/sspl:147]: Yes, Not required elif condition. Updated in the latest patch

**Review by ajaysrivas — APPROVED** [review]

File: ha/resource/sspl [hunk 4]
```diff
@@ -115,22 +115,48 @@ stateful_check_state() {
...
+                return 1
+            elif [[ $status == "unknown" ]]; then
+                ((count++))
+                if [ $count == 5 ]; then
+                    return 1
+                fi
+            fi
+            sleep 1
+        done
+}
```
Comment (ujjwalpl) [inline ha/resource/sspl:137]: No return value at the end. Caller is looking for a return value.

File: ha/resource/sspl [hunk 5]
```diff
@@ -115,22 +115,48 @@ stateful_check_state() {
     return 7
 }
 
+start_sspl_service() {
+        systemctl start sspl-ll
+        echo "START (master): $?"
```
Comment (ujjwalpl) [inline ha/resource/sspl:120]: Code should be written this way for better readability ... 

```
start_sspl_service() {
	echo "START (master): $?"
	systemctl start sspl-ll
    
    # check status
    for retry in {1..5}; do 
    	status=$(systemctl is-active sspl-ll)
    	case $status in
    		active )
    			return 0
    			;;
    			
    		failed )
    			return 1
    			;;
    			
    		* )
    			;;
    	esac
    	sleep 1
    done
    return 1
}
```

File: ha/resource/sspl [hunk 6]
```diff
@@ -115,22 +115,48 @@ stateful_check_state() {
...
+
 stateful_start() {
     stateful_check_state "master"
     if [ $? -eq 0 ]; then
         # CRM Error - Should never happen
         log "START (master)..."
-        systemctl start sspl-ll
-        log "START (master): $?"
-        return $OCF_RUNNING_MASTER
+        start_sspl_service
```
Comment (ujjwalpl) [inline ha/resource/sspl:144]:
```
start_sspl_service && return $OCF_RUNNING_MASTER
return $OCF_FAILED_MASTER
```

File: ha/resource/sspl (hunk 5 above)
Comment (VijayThakkar132) [reply ha/resource/sspl:120]: Raised below PR to address the above comment.
https://github.com/Seagate/cortx-ha/pull/296

File: ha/resource/sspl (hunk 6 above)
Comment (VijayThakkar132) [reply ha/resource/sspl:144]: Raised below PR to address the above comment.
https://github.com/Seagate/cortx-ha/pull/296

File: ha/resource/sspl (hunk 4 above)
Comment (VijayThakkar132) [reply ha/resource/sspl:137]: Raised below PR to address the above comment.
https://github.com/Seagate/cortx-ha/pull/296

**Pull request merged**